src/ILCompiler.WebAssembly/src/CodeGen/ILToWebAssemblyImporter.cs

```diff
@@ -1474,21 +1474,22 @@ private void ImportCasting(ILOpcode opcode, int token)
             
             //TODO: call GetCastingHelperNameForType from JitHelper.cs (needs refactoring)
             string function;
-            bool throwing = opcode == ILOpcode.castclass;
+            bool isCastClass = opcode == ILOpcode.castclass;
             if (type.IsArray)
-                function = throwing ? "CheckCastArray" : "IsInstanceOfArray";
+                function = isCastClass ? "CheckCastArray" : "IsInstanceOfArray";
             else if (type.IsInterface)
-                function = throwing ? "CheckCastInterface" : "IsInstanceOfInterface";
+                function = isCastClass ? "CheckCastInterface" : "IsInstanceOfInterface";
             else
-                function = throwing ? "CheckCastClass" : "IsInstanceOfClass";
+                function = isCastClass ? "CheckCastClass" : "IsInstanceOfClass";
 
             var arguments = new StackEntry[]
             {
                 new LoadExpressionEntry(StackValueKind.ValueType, "eeType", GetEETypePointerForTypeDesc(type, true), _compilation.TypeSystemContext.SystemModule.GetKnownType("System", "EETypePtr")),
                 _stack.Pop()
             };
 
-            _stack.Push(CallRuntime(_compilation.TypeSystemContext, TypeCast, function, arguments, type));
+            // don't force the return type fo isinst.  If the type is a struct and the argument a boxed struct, then it would not be correct.
```

Comment:
I think both castclass and ininst should be treated the same way here.

Reply:
I'm not sure how to do this correctly.  I want the correct type for the evaluation stack expression for value types (so we can tell later how to compare to `0`).  So you can see that for now I've used `MakePointerType` but dont think that's right either.

Reply:
isinst and castclass always return object reference that you can then compare to 0.

Reply:
Have changed to Object.